### ISSUE TITLE
bed exploit to bypass server/NPC fly checks

### ISSUE THREAD

**Issue by ghost** [PaperMC/Paper#399]
This issue has been moved to a private tracker and forwarded to the appropriate parties for resolution.

**Comment by Chnkr** [comment]
Sounds like an NCP issue to be honest.

**Comment by Chnkr** [comment]
This happens when `mc.thePlayer.sleeping` is set to `false` and `mc.thePlayer.sleepTimer` is set to 0 can easily reproduce it by setting those values when entering a bed using the latest paper build.

**Comment by Chnkr** [comment]
This is how i implemented it, works flawlessly.

```
public class HolyFuckPaperWTF extends Module
{
    public HolyFuckPaperWTF() {
        super("HolyFuckPaperWTF", 0, Category.Exploit, "???", Color.RED);
    }
    
    @Override
    public void onLivingUpdate() {
        if (!this.isToggled()) {
            return;
        }
        if (!this.mc.thePlayer.isPlayerSleeping()) {
            return;
        }
        this.mc.thePlayer.sleeping = false;
        this.mc.thePlayer.sleepTimer = 0;
        super.onLivingUpdate();
    }
}
```

**Comment by bonces** [comment]
https://www.youtube.com/watch?v=xWEJgz_hqbg

it's not just NCP, server completely stops checking for them, not even server.properties allow-flight=false working

**Comment by bonces** [comment]
only works in paper

**Comment by Chnkr** [comment]
This exploit can also be used to teleport great distances as @bonces mentioned before, this is done by placing a bed in the nether at the exact same co-ordinates your original bed was in.

Once you enter the nether portal with `sleeping` set to false whilst you are still sleeping on the server side, the server will simply send you to the bed you placed down earlier.

**Comment by Spottedleaf** [comment]
I can confirm this works on LATEST paper & LATEST spigot.

I reproduced by simply commenting out "this.sleeping = true" in EntityPlayer#trySleep in MCP (effectively making the client "think" as if it's not sleeping).

I also set my gamemode to spectator CLIENT-SIDE and enabled fly CLIENT-SIDE to test noclipping. Both were successful on PAPER and SPIGOT.

I confirmed this with another client.

However, player movement was still checked SERVER-SIDE, as attempting to set my velocity to (0, 100, 0) resulted in the server printing this message:
Player350 moved too quickly! 0.0,100.00000000000001,0.0
(BOTH SPIGOT & PAPER)

I have NOT confirmed teleportation with beds in the nether, as I have not tested that.

As I am writing this message, I was simply floating in the air WITHOUT movement as Player350, the server (SPIGOT) printed this message approximately 5 minutes after the video:
Player350 moved wrongly!

and immediately the client was woke up from the bed. I have no idea what happened and why it was so delayed.

VIDEO:
(same result for paper & spigot)

https://www.youtube.com/watch?v=c2ZUIkTHlQI
